### ISSUE TITLE
GLTF loader loses information about heirarchy

### ISSUE THREAD

**Issue by jpetkau** [bevyengine/bevy#4539]
`bevy_gltf::Gltf` should store names in a way that makes it possible (and hopefully easy) to reconstruct the heirarchy, including names, that was in the gltf json.

Currently, doing anything involving names inside a Gltf scene is very difficult. E.g. if I want to use some special name to mark whether a node is for collisions or for rendering or both, I can't.

Currently, `bevy_gltf::Gltf` stores information in a way makes it impossible to reconstruct the node hierarchy with node names. Names are retained only in a mapping from `name->Handle<GltfNode>`, but the heirarchy is stored as copies of child GltfNodes, so their identites are lost:

    pub struct GltfNode {
        pub children: Vec<GltfNode>,
        pub mesh: Option<Handle<GltfMesh>>,
        pub transform: bevy_transform::prelude::Transform,
    }

A workaround is to first load the gltf json using the `gltf` library directly, extract whatever structure is needed, and then use indices to find the corresponding nodes and meshes in the `bevy_gltf::Gltf` data.

Also, it *is* possible to reconstruct the *mesh* names, since meshes are stored as handles in the `GltfNode` heirarchy.

The simplest solution would be to just add `pub name: Option<String>` to `GltfNode`.

Probably better would be to reorganize `struct Gltf` a little bit to avoid the redundancy. E.g. change `children` to a vec of indices instead of copies of the `GltfNode` struct, and generally stick closer to the index-based structure in the gltf file (which is easier to work with than the bevy_gltf transformation of it.). The created `Scene` objects could also get `Name` components where appropriate.

(I'm happy to do make this change unless there's some objection).

**Comment by nicopap** [comment]
The GLTF loader does two things:
1. Create that `Gtlf` struct
2. Populate the ECS with the equivalent `Gltf` nodes, but with real `Transform` components etc.

Usually, you would use the gltf loader to get your things in the ECS and then use the `Name` component to find the entities you want and add to those entities your own custom components that do not exist in a gltf file.

Using names to locate entity nodes loaded from GLTF and customizing them is a known pattern, also explored by @Sheepyhead

We've been using the very same pattern in Warlock's Gambit (you might be familiar with it :wink:). The [bevy-scene-hook](https://github.com/nicopap/bevy-scene-hook) crate isolates the relevant code from the game and demonstrates this pattern as simple re-usable code.

**Comment by Sheepyhead** [comment]
> Using names to locate entity nodes loaded from GLTF and customizing them is a known pattern, also explored by @Sheepyhead

I was actually going to explore using the gltf extras field for this purpose now that we add that with the gltf load, as matching by name has a few issues:
1. The Name component can be added by anything, thus you can unknowingly run into one of the hooks you've made without ever loading a gltf file and
2. Blender doesn't allow for multiple objects of the same name. Afaik this isn't a limitation with gltf but it's still a compatibility issue

**Comment by jpetkau** [comment]
@nicopap the problem is that the `bevy_gltf` loader actually discards the information about which names go where in the heirarchy. It's no longer available from either the `Gltf` struct or the ECS version of the scene.

(I guess strictly speaking you *could* reconstruct it from the ECS info: it gives them a Name component with names like `GltfNode3`. You could parse that to recover the node index, then index into in the `nodes` field of `Gltf` to get the handle, then scan backwards through the `named_nodes` map to figure out the corresponding name. But ain't nobody got time for that.)

Ideally the `Name` component in a scene would contain the actual exported name, not `GltfNode3` or whatever.

I also found it's difficult to work with the scene in ECS format, if I want to do anything other than simple instantiation. Sometimes a plain old data structure what you want.

(Also, unrelated: sorry for ghosting everyone in the jam :-( )

**Comment by mockersf** [comment]
> Ideally the `Name` component in a scene would contain the actual exported name, not `GltfNode3` or whatever.

Those names should only be generated when there isn't a name on the node. Do you have an example gltf file where that is not the case?

**Comment by Sheepyhead** [comment]
> > (1) The Name component can be added by anything, thus you can unknowingly run into one of the hooks you've made without ever loading a gltf file, and (2) Blender doesn't allow for multiple objects of the same name.
> > @Sheepyhead
> 
> If you had a rule that just looked for any new entity with a particular `Name`, that could cause problems. But a rule that is applied in game-specific object instantiation code can do whatever it needs to. Likewise the unique name requirement in Blender isn't a problem as long as whatever naming convention the game uses takes that into account. (E.g. the convention that collision meshes are all named "coll" wouldn't work, but the convention that they start with "coll_" works fine.)

I agree it's possible to work around this issue, but I don't think you should have to, and having a dedicated field in extras seems like the more sane choice rather than ascribing meaning to specific names, or even worse, specific affixes.

**Comment by jpetkau** [comment]
> I agree it's possible to work around this issue, but I don't think you should have to, and having a dedicated field in extras seems like the more sane choice rather than ascribing meaning to specific names, or even worse, specific affixes.

? I'm not sure what "work around" you're referring to. I'm not trying to do anything with the name of entities already in the game; I'm trying to do something with names in the GLTF file as I instantiate new ones. And the naming convention in a GLTF file is whatever some particular game wants it to be.

(That said, I certainly have no objection to supporting extras field as well. But 'name' exists GLTF for the purpose of identifying particular nodes by name, and I think that's a perfectly reasonable thing to want to use it for.)

**Comment by jpetkau** [comment]
> > Ideally the `Name` component in a scene would contain the actual exported name, not `GltfNode3` or whatever.
> 
> Those names should only be generated when there isn't a name on the node. Do you have an example gltf file where that is not the case?

Oops, you are correct. I misread the code.

That means that the heirarchy information I'm talking about *is* available via `Scene`, just not via `GltfNode`. So it *is* possible to do what I want, just less conveniently. (However, it would still be helpful if the heirarchy in `Gltf.nodes` file simply kept the names in the first place.)

**Comment by nicopap** [comment]
Maybe the `Entity` corresponding to the `GltfNode` should be added to the `Gltf` struct nodes? This seems like something natural. Though personally, once I found out I could manipulate the scene within the ECS world rather than through the `Gltf` asset, I never looked back into it.

**Comment by jpetkau** [comment]
> Though personally, once I found out I could manipulate the scene within the ECS world rather than through the `Gltf` asset, I never looked back into it.

Yeah, I think the "right" way to do what I want is to just ignore `struct Gltf` and instead use the ECS components in the `Scene`, especially since it turns out the names *are* preserved correctly there.

The reason I didn't do it that way in the first place is the general problem with ECS designs: it's hard to figure out what's in there. I.e. it's very easy to examine the source of `struct Gltf` and figure out exactly what is and isn't present; it's very difficult to do the same with the `Scene` version. But once you're past that documentation hurdle, working with the `Scene` is fine.

I'll still send the PR to add names to `GltfNode`, since if the struct is there at all there's no reason it should be missing info like that.

But maybe a better design would actually replace most of `struct Glt` with handles into the scene, and replace all the public data with utility methods that query the World for you? (That would avoid some redundancy, *and* serve as an example of what's inside the Scene object and how to access it.)

**Comment by Sheepyhead** [comment]
> ? I'm not sure what work around you're referring to. I'm not trying to do anything with the name of entities already in the game; I'm trying to do something with names in the GLTF file as I instantiate new ones. There's no duplicate name problem to work around.

That is in fact the problem. Any entity can have a Name component, that's not unique to things you just imported from a gltf file. The extras field is less limited and also explicitly related to imported GLTF models. The workaround I was referring to was your affix idea, which is a workaround, but not a very elegant one. 


In any case I wasn't saying what people should or shouldn't do, just commenting on what I consider best practice, and what I'd be working on for my scene loading example next.

**Comment by omer0909** [comment]
``` Rust
#[derive(Component, Reflect)]
#[reflect(Component)]
struct AddCollider;

#[derive(Default, Clone)]
struct CollisionExtension;

impl GltfExtensionHandler for CollisionExtension {
    fn dyn_clone(&self) -> Box<dyn GltfExtensionHandler> {
        Box::new(self.clone())
    }

    fn on_spawn_mesh_and_material(
        &mut self,
        load_context: &mut LoadContext<'_>,
        primitive: &gltf::Primitive,
        mesh: &gltf::Mesh,
        material: &gltf::Material,
        entity: &mut EntityWorldMut,
    ) {
        if let Some(name) = entity.get::<Name>() {
            println!("Object Name: {}", name);
        }

        let mesh_name = mesh.name().unwrap_or("Unnamed Mesh");
        println!("Mesh Name: {}", mesh_name);
        println!("--------------------------------");
        entity.insert(AddCollider);
    }
}
```

output:
```
Object Name: Plane
Mesh Name: Plane
--------------------------------
Object Name: Suzanne
Mesh Name: Suzanne
--------------------------------
Object Name: Suzanne
Mesh Name: Suzanne
--------------------------------
```

[scene.zip](https://github.com/user-attachments/files/25461506/scene.zip)

The output should be as follows, based on the ".glb" file from the blender:
```
Object Name: Ground
Mesh Name: Plane
--------------------------------
Object Name: Monkey1
Mesh Name: Suzanne
--------------------------------
Object Name: Monkey2
Mesh Name: Suzanne
--------------------------------
```
So when importing objects, it imports them based on the mesh name inside the object rather than the object name. This is a problem because many objects can use the same mesh. (Instancing)
